### PR TITLE
[Bug fix] Enable UnpackToDestFp32 in matmul when appropriate

### LINKED ISSUE #46101: [matmul]: PCC Of `ttnn.matmul` Drops Regardless Of `fp32_dest_acc_en` Value For BFloat16 Inputs With High Inner Dim
### Component / Area

ops

### Issue Type (optional)

Bad Outputs

### Observed

Full analysis of `ttnn.matmul` through Forge Compiler:

## PCC Analysis

```
Llama 3.2 1b - vocab_size = 128256 (0 steps)

%1551 = "ttnn.matmul"(%1550, %arg281) <{compute_config = #ttnn.device_compute_kernel_config<math_fidelity = hifi4, fp32_dest_acc_en = true>, transpose_a = false, transpose_b = false}> : (tensor<256x128256xbf16, #ttnn_layout95>, tensor<128256x2048xbf16, #ttnn_layout63>) -> tensor<256x2048xbf16, #ttnn_layout74> loc(#loc1299)

- fp32_dest_acc_en on: 0.975314
- fp32_dest_acc_en off: 0.975314

Gemma 1.1 2b it - vocab_size 256000 (200 steps)

%1703 = "ttnn.matmul"(%1702, %arg315) <{compute_config = #ttnn.device_compute_kernel_config<math_fidelity = hifi4, fp32_dest_acc_en = true>, transpose_a = false, transpose_b = false}> : (tensor<256x256000xbf16, #ttnn_layout82>, tensor<256000x2048xbf16, #ttnn_layout55>) -> tensor<256x2048xbf16, #ttnn_layout52> loc(#loc1463)

- fp32_dest_acc_en on: 0.841361
- fp32_dest_acc_en off: > 0.99
```

From these results we see that failure is being propagated more and more as we go in vocab_size. What is interesting is that when we turn of `fp32_dest_acc_en` flag in `Gemma 1.1 2b it` we get good PCC in chisel for 200 step checkpoint, but after letting the training go for a little longer, we experience problems in later steps.

## Loss Curve Analysis

What is important to look at here is the loss curve:
<table>
  <tr>
    <td align="center">
      Llama 3.2 1b
    </td>
    <td align="center">
Image
    </td>
  </tr>
  <tr>
    <td align="center"><strong>Llama 3.2 1b</strong></td>
    <td align="center"><strong>Gemma 1.1 2b it</strong></td>
  </tr>
</table>

We can see that for Gemma, even after training with `fp32_dest_acc_en` off, we get divergence in loss curve. The one thing is definitive, **ONLY** `fp32_dest_acc_en` + `packer_l1_acc` knobs together produce a valid training that does not diverge and perfectly follows CPU baseline.

### Expected

External user should not be worried about matmuls not working properly for large inner dim just because they did not activate special-case flags like `packer_l1_acc` or `fp32_dest_acc_en`.  
From https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/op_kernel_dev/accuracy_tips/accuracy_tips.md we can see that the problem is highly likely related to `unpack_to_dest_mode`, which should not be something user has to worry about.

Also documentation in the repo still suggests that using `packer_l1_acc` with `fp32_dest_acc_en` is not recommended. https://github.com/tenstorrent/tt-metal/blob/main/docs/L1_ACCUMULATION_FP32_ANALYSIS.md

### 1. Steps (exact commands)

Use matmul on provided inputs (or really small inputs) where `fp32_dest_acc_en` is set to true.

[repro.py](https://github.com/user-attachments/files/28603603/repro.py)

### 2. Input data / link or description

```
torch.manual_seed(seed)
scale = 0.001
in_lhs = (torch.randn(32, k) * scale).bfloat16()
in_rhs = (torch.randn(k, 32) * scale).bfloat16()
return in_lhs, in_rhs
```

k should be inner-dim of at least 100k

### 3. Frequency

Always, when inputs are small.

### 1. Software Versions

Ubuntu 24.04

### 2. Hardware Details

All products experience this.

### Is this a regression?

Unknown

### Regression Details

_No response_

### Logs & Diagnostics

_No response_

### Priority

None

### Impact

_No response_

### PR DESCRIPTION
**Pull request by fplavecTT**
### Summary
#46101 reported low PCC of matmul with some combinations of `fp32_dest_acc_en` and `packer_l1_acc`.
While `fp32_dest_acc_en=False` results are expected, it uncovered that there was an unexpected drop in PCC for the `fp32_dest_acc_en=True` and `packer_l1_acc=False` combination, because matmuls were not correctly setting the `UnpackToDestFp32` flag for that scenario.
This PR corrects this so that this flag is set on relevant 32-bit CBs when `fp32_dest_acc_en=True`.
For CBs that are read both through Src and Dest, an alias CB (pointing to the same SRAM) is added so that each can have `unpack_to_dest_mode` set appropriately.
It also adds targeted tests to prove the issue (fails before, passes after the fix) and prevent future regressions.
Some code paths could only be exercised by T3K tests, so the tests were added there.

With these changes, running the `repro.py` script from #46101 shows significant accuracy improvement for the `fp32_dest_acc_en=True`, `packer_l1_acc=False` combination, as expected:

Before:
| fp32_dest_acc_en | packer_l1_acc | PCC_vs_bf16 | PCC_vs_fp32 | max\|bf16\| | max\|fp32\| |
|:---:|:---:|:---:|:---:|:---:|:---:|
| False | False | 0.508860 | 0.508850 | 5.3369e-01 | 5.3371e-01 |
| False | True  | 0.997214 | 0.997223 | 2.0508e-02 | 2.0330e-02 |
| **True**  | **False** | **0.922274** | **0.922256** | **8.5938e-02** | **8.6116e-02** |
| True  | True  | 0.999999 | 0.999999 | 4.8828e-04 | 4.4930e-04 |

After: 
| fp32_dest_acc_en | packer_l1_acc | PCC_vs_bf16 | PCC_vs_fp32 | max\|bf16\| | max\|fp32\| |
|:---:|:---:|:---:|:---:|:---:|:---:|
| False | False | 0.508860 | 0.508850 | 5.3369e-01 | 5.3371e-01 |
| False | True  | 0.997214 | 0.997223 | 2.0508e-02 | 2.0330e-02 |
| **True**  | **False** | **0.999999** | **0.999999** | **4.8828e-04** | **4.4930e-04** |
| True  | True  | 1.000000 | 0.999999 | 4.8828e-04 | 4.4930e-04 |

Closes #46101

### Testing
- New T3K tests are failing before: https://github.com/tenstorrent/tt-metal/actions/runs/29537061082 and passing after the fix: https://github.com/tenstorrent/tt-metal/actions/runs/29537252093 
- The fix caused a performance regression in `test_device_perf_pdl_110_cores[test_panoptic_deeplab_110_cores]` and `test_device_perf_pdl_110_cores[test_panoptic_deeplab_110_cores]`: https://github.com/tenstorrent/tt-metal/actions/runs/29537531899/job/87797544726
These use matmul through convolution operations. Setting `fp32_dest_acc_en=False` did not work (PCC drop was huge). Instead, the model was adjusted to set `packer_l1_acc=True`, which is the other way to achieve 32-bit accumulation in matmul. However, this option comes at a cost of SRAM, so it cannot be applied to all convolutions; instead it was applied selectively to convolutions with most significant perf impact and where SRAM cost is modest so the model still fits.
With these changes, panoptic passes: https://github.com/tenstorrent/tt-metal/actions/runs/29626476495/job/88032702781

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:fplavec/46101_matmul_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:fplavec/46101_matmul_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-sanity-tests.yaml?query=branch:fplavec/46101_matmul_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:fplavec/46101_matmul_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:fplavec/46101_matmul_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:fplavec/46101_matmul_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:fplavec/46101_matmul_fixes)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:fplavec/46101_matmul_fixes)

### CODE REVIEW

File: tests/ttnn/unit_tests/operations/matmul/test_matmul.py [hunk 1]
```diff
@@ -4010,3 +4010,199 @@ def test_matmul_kt_not_divisible_by_in0_block_w_rejected(device, expect_error):
...
+        frobenius_threshold=0.2,
+        check_allclose=False,
+        check_pcc=True,
+        check_frobenius=True,
+        check_ulp=False,
+    )
+
+
+@pytest.mark.xfail(
+    reason=("Fused matmul untilize_out loses fp32 precision. Issue #49836"),
```
[the remark this answers is not in the document]
Comment (fplavecTT) [reply tests/ttnn/unit_tests/operations/matmul/test_matmul.py:4149]: > Codebase convention: Include the full issue URL

@copilot Where did you find this convention? What if the URL changes in the future?

[the remark this answers is not in the document]
Comment (fplavecTT) [reply tests/ttnn/unit_tests/operations/matmul/test_matmul.py:4149]: @copilot Did you compare how many instances of this convention exist compared to #number convention?
I asked another agent to do the comparison and it said this:
> Bare #NNNN: ~174
> Full GitHub issue URL: ~12

According to that, convention would be the opposite of what you claim

**Review by mbezuljTT — APPROVED** [review]

File: tests/nightly/t3000/ccl/test_minimal_all_gather_matmul_async.py [hunk 2]
```diff
@@ -500,3 +549,120 @@ def test_all_gather_matmul_async(
...
+        enable_trace=False,
+        use_barrier=False,
+        use_persistent_buffers=False,
+        packer_l1_acc=False,
+        precision_offset=1000.0,
+    )
+
+
+# The mcast_in1 path segfaults for this single-core non-gather 1D matmul routed through
+# all_gather_matmul, and reproduces on main. A segfault would kills the test process; so run=False
```
[the remark this answers is not in the document]
Comment (fplavecTT) [reply tests/nightly/t3000/ccl/test_minimal_all_gather_matmul_async.py]: Fixed

**Review by williamlyTT — APPROVED** [review]

File: tests/nightly/t3000/ccl/test_minimal_all_gather_matmul_async.py [hunk 3]
```diff
@@ -166,28 +181,56 @@ def run_all_gather_impl(
...
         fp32_dest_acc_en=True,
-        packer_l1_acc=True,
+        packer_l1_acc=packer_l1_acc,
     )
 
     ##### Perform torch ops #####
     torch_matmul_output_list = []
     for i in range(num_iters):
-        if use_bias:
+        if use_bias and precision_offset is not None:
```
Comment (rmillerTT) [inline tests/nightly/t3000/ccl/test_minimal_all_gather_matmul_async.py]: I would prefer:

```python
if use_bias:
    g = ag_output_tensor_goldens_list[i]
    w = weights_tensor.T.contiguous()
    b = bias_tensor

    if precision_offset is not None:
        matmul_output = torch.nn.functional.linear(g.double(), w.double(), b.double()).float()
    else:
        matmul_output = torch.nn.functional.linear(g, w, b)
```

File: ttnn/cpp/ttnn/operations/matmul/device/kernels/compute/bmm_large_block_zm_fused_bias_activation.cpp [hunk 4]
```diff
@@ -317,6 +340,7 @@ void kernel_main() {
                                     in0_cb_id,
                                     in1_cb_id,
                                     mm_partials_cb_id,
+                                    mm_partials_reload_cb_id,
```
Comment (rmillerTT) [inline ttnn/cpp/ttnn/operations/matmul/device/kernels/compute/bmm_large_block_zm_fused_bias_activation.cpp:343]: I'd change this to mm_partials_alias_cb_id to more tightly link it to the preceding variable

**Review by rmillerTT — APPROVED** [review]

File: ttnn/cpp/ttnn/operations/matmul/device/kernels/compute/bmm_large_block_zm_fused_bias_activation.cpp (hunk 4 above)
Comment (fplavecTT) [reply ttnn/cpp/ttnn/operations/matmul/device/kernels/compute/bmm_large_block_zm_fused_bias_activation.cpp:343]: This is not always an alias (see comments on lines 204-205), so that would be misleading

File: tests/nightly/t3000/ccl/test_minimal_all_gather_matmul_async.py (hunk 3 above)
Comment (fplavecTT) [reply tests/nightly/t3000/ccl/test_minimal_all_gather_matmul_async.py]: Yes, that's cleaner. Done.

**Review by jvegaTT — APPROVED** [review]
